src/NuGet.Clients/NuGet.PackageManagement.UI/UserInterfaceService/NuGetSourcesServiceWrapper.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Configuration;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.UI
+{
+    internal sealed class NuGetSourcesServiceWrapper : INuGetSourcesService
+    {
+        private INuGetSourcesService _service = NullNuGetSourcesService.Instance;
+        private readonly object _syncObject = new object();
+
+        public event EventHandler<IReadOnlyList<PackageSourceContextInfo>>? PackageSourcesChanged;
+
+        internal INuGetSourcesService Service
+        {
+            get
+            {
+                lock (_syncObject)
+                {
+                    return _service;
+                }
+            }
+        }
+
+        public INuGetSourcesService Swap(INuGetSourcesService newService)
+        {
+            lock (_syncObject)
+            {
+                Service.PackageSourcesChanged -= OnPackageSourcesChanged;
+
+                INuGetSourcesService oldService = _service;
+
+                _service = newService ?? NullNuGetSourcesService.Instance;
+
+                Service.PackageSourcesChanged += OnPackageSourcesChanged;
+
+                return oldService;
+            }
+        }
+
+        private void OnPackageSourcesChanged(object sender, IReadOnlyList<PackageSourceContextInfo> e)
+        {
+            PackageSourcesChanged?.Invoke(sender, e);
+        }
+
+        public void Dispose()
+        {
+            using (INuGetSourcesService? service = _service)
+            {
+                _service = NullNuGetSourcesService.Instance;
+            }
```

Comment:
I think this should be:
```C#
lock (_syncObject)
{
    UnregisterEventHandlers();

    using (INuGetSolutionManagerService? service = _service)
    {
        _service = NullNuGetSolutionManagerService.Instance;
    }
}
```
This feedback applies to the other service wrapper (`NuGetSolutionManagerServiceWrapper`) too.